overlord/snapstate/backend/setup.go

```diff
@@ -48,43 +56,46 @@ func (b Backend) SetupSnap(snapFilePath, instanceName string, sideInfo *snap.Sid
 		if err == nil {
 			return
 		}
-		// XXX: this will also remove the snap from /var/lib/snapd/snaps
-		if e := b.RemoveSnapFiles(s, s.GetType(), meter); e != nil {
+
+		// this may remove the snap from /var/lib/snapd/snaps depending on undoCtx
+		if e := b.RemoveSnapFiles(s, s.GetType(), undoCtx, meter); e != nil {
 			meter.Notify(fmt.Sprintf("while trying to clean up due to previous failure: %v", e))
 		}
 	}()
 
 	if err := os.MkdirAll(instdir, 0755); err != nil {
-		return snapType, err
+		return snapType, nil, err
 	}
 
 	if s.InstanceKey != "" {
 		err := os.MkdirAll(snap.BaseDir(s.SnapName()), 0755)
 		if err != nil && !os.IsExist(err) {
-			return snapType, err
+			return snapType, nil, err
 		}
 	}
 
-	if err := snapf.Install(s.MountFile(), instdir); err != nil {
-		return snapType, err
+	var nothingToDo bool
+	if nothingToDo, err = snapf.Install(s.MountFile(), instdir); err != nil {
```

Comment:
this is more "didNothing", than "nothing to do", is not about us, it's about what Install did or  didn't

Reply:
Indeed, renamed.